src/time_integration.jl

```diff
@@ -1,4 +1,665 @@
+                    copyto!(η0,η1)
+                end
+            end
+        end
+
+        ghost_points!(u0,v0,η0,S)
+
+        # type conversion for mixed precision
+        u0rhs = convert(Diag.PrognosticVarsRHS.u,u0)
+        v0rhs = convert(Diag.PrognosticVarsRHS.v,v0)
+        η0rhs = convert(Diag.PrognosticVarsRHS.η,η0)
+
+        # ADVECTION and CORIOLIS TERMS
+        # although included in the tendency of every RK substep,
+        # only update every nstep_advcor steps if nstep_advcor > 0
+        if dynamics == "nonlinear" && nstep_advcor > 0 && (i % nstep_advcor) == 0
+            UVfluxes!(u0rhs,v0rhs,η0rhs,Diag,S)
+            advection_coriolis!(u0rhs,v0rhs,η0rhs,Diag,S)
+        end
+
+        # DIFFUSIVE TERMS - SEMI-IMPLICIT EULER
+        # use u0 = u^(n+1) to evaluate tendencies, add to u0 = u^n + rhs
+        # evaluate only every nstep_diff time steps
+        if (i % nstep_diff) == 0
+            bottom_drag!(u0rhs,v0rhs,η0rhs,Diag,S)
+            diffusion!(u0rhs,v0rhs,Diag,S)
+            add_drag_diff_tendencies!(u0,v0,Diag,S)
+            ghost_points_uv!(u0,v0,S)
+        end
+
+        t += dtint
+
+        # TRACER ADVECTION
+        u0rhs = convert(Diag.PrognosticVarsRHS.u,u0)  # copy back as add_drag_diff_tendencies changed u0,v0
+        v0rhs = convert(Diag.PrognosticVarsRHS.v,v0)
+        tracer!(i,u0rhs,v0rhs,Prog,Diag,S)
+
+        # feedback and output
+        feedback.i = i
+        feedback!(Prog,feedback,S)
+        output_nc!(i,netCDFfiles,Prog,Diag,S)       # uses u0,v0,η0
+
+        if feedback.nans_detected
+            break
+        end
+
+        # Copy back from substeps
+        copyto!(u,u0)
+        copyto!(v,v0)
+        copyto!(η,η0)
+    end
+
+    # finalise feedback and output
+    feedback_end!(feedback)
+    output_close!(netCDFfiles,feedback,S)
+
+    if S.parameters.return_time
+        return feedback.tend - feedback.t0
+    else
+        return PrognosticVars{Tprog}(remove_halo(u,v,η,sst,S)...)
+    end
+end
+
+"""Integrate ShallowWaters forward and checkpoints the integration. This is the
+integration function needed if we want to use Enzyme on the model."""
+function checkpointed_integration(S, scheme)
+
+    # setup
+    Diag = S.Diag
+    Prog = S.Prog
+
+    @unpack u,v,η,sst = Prog
+    @unpack u0,v0,η0 = Diag.RungeKutta
+    @unpack u1,v1,η1 = Diag.RungeKutta
+    @unpack du,dv,dη = Diag.Tendencies
+    @unpack du_sum,dv_sum,dη_sum = Diag.Tendencies
+    @unpack du_comp,dv_comp,dη_comp = Diag.Tendencies
+
+    @unpack um,vm = Diag.SemiLagrange
+
+    @unpack dynamics,RKo,RKs,tracer_advection = S.parameters
+    @unpack time_scheme,compensated = S.parameters
+    @unpack RKaΔt,RKbΔt = S.constants
+    @unpack Δt_Δ,Δt_Δs = S.constants
+
+    @unpack nt,dtint = S.grid
+    @unpack nstep_advcor,nstep_diff,nadvstep,nadvstep_half = S.grid
+
+    # calculate layer thicknesses for initial conditions
+    ShallowWaters.thickness!(Diag.VolumeFluxes.h,η,S.forcing.H)
```

Comment:
Why is the `ShallowWaters.` necessary here? Are we not in the same module still?

Reply:
It's not, this was a cut-and-paste from different code and I forgot to delete it